File: 백준/Gold/16940. BFS 스페셜 저지/BFS 스페셜 저지.py
from collections import deque

def check():
    for i in range(1,N+1):
        if sequences[i] != result[i-1]:
            return False

    return True

if __name__ == "__main__":
    N = int(input())
    adj = [[] for _ in range(N+1)]

    for _ in range(N-1):
        u, v = map(int,input().split())
        adj[u].append(v)
        adj[v].append(u)

    sequences = [0] + list(map(int,input().split()))
    order = [ 0 for _ in range(N+1)]
    for i in range(1,N+1):
        order[sequences[i]] = i

    for i in range(1,N+1):
        adj[i].sort(key = lambda idx : order[idx])


    visited = [False for _ in range(N+1)]
    result = []

    q = deque([1])
    visited[1] = True
    result.append(1)
    while q:
        cur = q.popleft()

        for nxt in adj[cur]:
            if visited[nxt]:
                continue

            visited[nxt] = True
            result.append(nxt)
            q.append(nxt)
    print( 1 if check() else 0)

File: 백준/Gold/16940. BFS 스페셜 저지/BFS 스페셜 저지.java
import java.util.*;
import java.io.*;

public class Main{
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		
		sequence = new int[N+1];
		order = new int[N+1];
		adj = new List[N+1];
		for(int i = 1; i<=N; i++) {
			adj[i] = new ArrayList<>();
		}
		
		for(int i = 0; i < N - 1; i++) {
			st = new StringTokenizer(br.readLine());
			int from = Integer.parseInt(st.nextToken());
			int to = Integer.parseInt(st.nextToken());
			adj[from].add(to);
			adj[to].add(from);
		}
		
		st = new StringTokenizer(br.readLine());
		for(int i = 1; i<=N; i++) {
			sequence[i] = Integer.parseInt(st.nextToken());
		}
		
		for(int i = 1;i <=N;i++) {
			order[sequence[i]] = i;
		}
		
		for(int i = 1; i<=N; i++) {
			Collections.sort(adj[i], (a,b)->{
				return order[a] - order[b];
			});
		}
		
		Queue<Integer> q = new ArrayDeque<>();
		visited = new boolean[N+1];
		q.add(1);
		visited[1] = true;
		result = new ArrayList<>();
		
		while(!q.isEmpty()) {
			int cur = q.poll();
			result.add(cur);
			
			for(int nxt : adj[cur]) {
				if(visited[nxt])
					continue;
				visited[nxt] = true;
				q.add(nxt);
			}
		}
		
		System.out.println(check() ? 1 : 0);
	}  
	static boolean check() {
		for(int i = 1; i<=N; i++) {
			if(sequence[i] != result.get(i-1))
				return false;
		}
		return true;
	}
	static List<Integer> result;
	static boolean visited[];
	static StringTokenizer st;
	static int N;
	static int sequence[], order[];
	static List<Integer> adj[];
}